### PR TITLE
Add methods to check workflow size violations for several other fields

### PR DESCRIPTION
**Pull request by MichaelSnowden**
**What changed?**
I generalized our workflow size limit constraints to 4 fields (child workflows, activities, external signals, and cancel requests).

**Why?**
I made this change so that we can limit these fields which could potentially cause issues if they're too large.

**How did you test it?**
I added unit tests.

**Potential risks**
These new methods aren't called yet.

**Is hotfix candidate?**
No.

### CODE REVIEW

File: service/history/commandChecker.go [hunk 1]
```diff
@@ -191,23 +197,62 @@ func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
...
+	c.metricsHandler.Counter(metricName).Record(1)
 	err := fmt.Errorf(
-		"the number of pending child workflow executions, %d, "+
+		"the number of %s, %d, "+
 			"has reached the error limit of %d established with %q",
+		resourceName,
 		numPending,
 		errLimit,
-		dynamicconfig.NumPendingChildExecutionsLimitError,
+		dynamicConfigKey,
```
Comment (yux0) [inline service/history/commandChecker.go:210]: I think this is too much. We don't need to error out the dynamicConfigKey.

File: service/history/commandChecker.go [hunk 2]
```diff
@@ -191,23 +197,62 @@ func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
...
-	numPending := len(c.mutableState.GetPendingChildExecutionInfos())
-	errLimit := c.numPendingChildExecutionsLimit
 	if withinLimit(numPending, errLimit) {
 		return nil
 	}
-	c.metricsHandler.Counter(metrics.NumPendingChildWorkflowsTooHigh.GetMetricName()).Record(1)
+	c.metricsHandler.Counter(metricName).Record(1)
 	err := fmt.Errorf(
-		"the number of pending child workflow executions, %d, "+
+		"the number of %s, %d, "+
```
Comment (yux0) [inline service/history/commandChecker.go:205]: I think we want to log this workflow id and run id?

File: service/history/commandChecker.go [hunk 3]
```diff
@@ -191,23 +197,62 @@ func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
...
+	)
+}
+
+func (c *workflowSizeChecker) checkIfNumPendingCancelRequestsExceedsLimit() error {
+	return c.checkCountConstraint(
+		len(c.mutableState.GetPendingRequestCancelExternalInfos()),
+		c.numPendingCancelsRequestLimit,
+		dynamicconfig.NumPendingCancelRequestsLimitError,
+		metrics.TooManyPendingCancelRequests.GetMetricName(),
+		"pending requests to cancel other workflows",
```
Comment (yux0) [inline service/history/commandChecker.go:242]: external workflow?

File: service/history/commandChecker.go [hunk 4]
```diff
@@ -191,23 +197,62 @@ func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
...
 	return err
 }
 
+func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
+	return c.checkCountConstraint(
+		len(c.mutableState.GetPendingChildExecutionInfos()),
+		c.numPendingChildExecutionsLimit,
+		dynamicconfig.NumPendingChildExecutionsLimitError,
+		metrics.TooManyPendingChildWorkflows.GetMetricName(),
+		"pending child workflow executions",
```
Comment (yux0) [inline service/history/commandChecker.go:222]: make those constants?

File: service/history/commandChecker.go [hunk 5]
```diff
@@ -191,23 +191,66 @@ func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
...
-	c.metricsHandler.Counter(metrics.NumPendingChildWorkflowsTooHigh.GetMetricName()).Record(1)
+	c.metricsHandler.Counter(metricName).Record(1)
 	err := fmt.Errorf(
-		"the number of pending child workflow executions, %d, "+
-			"has reached the error limit of %d established with %q",
+		"the number of %s, %d, for run %q of workflow %q in namespace %q has reached the error limit of %d",
+		resourceName,
 		numPending,
+		key.RunID,
+		key.WorkflowID,
```
Comment (yux0) [inline service/history/commandChecker.go:203]: I don't mean to have the workflow id and run id in the error. Can we log an error message? Or the caller will do it?

Comment (MichaelSnowden) [reply service/history/commandChecker.go:203]: We do that on line 207, and the logger is tagged with those values on 187

Comment (MichaelSnowden) [reply service/history/commandChecker.go]: I've removed this from the inline error message

File: service/history/commandChecker.go [hunk 6]
```diff
@@ -191,23 +191,63 @@ func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
...
 	err := fmt.Errorf(
-		"the number of pending child workflow executions, %d, "+
-			"has reached the error limit of %d established with %q",
+		"the number of %s, %d, has reached the per-workflow limit of %d",
+		resourceName,
 		numPending,
 		errLimit,
-		dynamicconfig.NumPendingChildExecutionsLimitError,
 	)
 	logger.Error(err.Error(), tag.Error(err))
```
Comment (yux0) [inline service/history/commandChecker.go:204]: nit: you can do
```
logger.Error("some error message", 
    tag.WorkflowNamespaceID(key.NamespaceID),
    tag.WorkflowID(key.WorkflowID),
    tag.WorkflowRunID(key.RunID),
    tag.Error(err))
```

File: service/history/commandChecker.go [hunk 7]
```diff
@@ -191,23 +191,63 @@ func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
...
 	return err
 }
 
+const (
```
Comment (yux0) [inline service/history/commandChecker.go:208]: nit: move const to the top?

**Review by yux0 — APPROVED** [review]

File: service/history/commandChecker.go (hunk 7 above)
Comment (MichaelSnowden) [reply service/history/commandChecker.go:208]: Will do in a follow up, keeping in the interest of time

File: service/history/commandChecker.go (hunk 6 above)
Comment (MichaelSnowden) [reply service/history/commandChecker.go:204]: Yeah, I did it this way because originally I logged a warning as well. I'll clean up in a follow up

**Pull request merged**